lib/hdb/common.c

```diff
@@ -98,6 +99,49 @@ hdb_value2entry_alias(krb5_context context, krb5_data *value,
     return decode_hdb_entry_alias(value->data, value->length, ent, NULL);
 }
 
+/*
+ * Some old databases may not have stored the salt with each key, which will
+ * break clients when aliases are used. Generate a default salt based on the
+ * real principal name in the entry to handle this case.
+ */
+static krb5_error_code
+add_default_salts(krb5_context context, HDB *db, hdb_entry *entry)
+{
+    krb5_error_code ret;
+    size_t i;
+    krb5_salt pwsalt;
+
+    ret = krb5_get_pw_salt(context, entry->principal, &pwsalt);
+    if (ret)
+	return ret;
+
+    for (i = 0; i < entry->keys.len; i++) {
+	Key *key = &entry->keys.val[i];
+
+	if (key->salt != NULL ||
```

Comment:
It might be worth noting that if an explicit key was set instead of deriving via `string_to_key()`, then `key->salt == NULL`.  However, the default salt would still be the same.  Its just that the salt would be ignored by the client when a keytab is used.  Generating the default salt and adding it to the key doesn't alter any behavior.